libfeed/feed-store-provider.c

```diff
@@ -0,0 +1,1239 @@
+                                                   cancellable,
+                                                   &local_error);
+
+  if (results == NULL)
+    {
+      g_task_return_error (task, g_steal_pointer (&local_error));
+      return;
+    }
+
+  g_task_return_pointer (task, g_steal_pointer (&results), (GDestroyNotify) g_ptr_array_unref);
+}
+
+static void
+append_stores_task_from_proxy (EosDiscoveryFeedKnowledgeAppProxy *ka_proxy,
+                               AppendStoresFromProxyFunc          proxy_func,
+                               gpointer                           proxy_func_data,
+                               GCancellable                      *cancellable,
+                               GAsyncReadyCallback                callback,
+                               gpointer                           user_data)
+{
+  g_autoptr(GTask) task = g_task_new (NULL, cancellable, callback, user_data);
+  g_autoptr(AppendStoresTaskData) data = append_stores_task_data_new (ka_proxy,
+                                                                      proxy_func,
+                                                                      proxy_func_data);
+
+  g_task_set_return_on_cancel (task, TRUE);
+  g_task_set_task_data (task,
+                        g_steal_pointer (&data),
+                        (GDestroyNotify) append_stores_task_data_free);
+  g_task_run_in_thread (task, append_stores_task_from_proxy_thread);
+}
+
+typedef struct _AllTasksResultsClosure
+{
+  GPtrArray      *results;
+  guint           remaining;
+  GTask          *task;
+} AllTasksResultsClosure;
+
+static AllTasksResultsClosure *
+all_tasks_results_closure_new (GDestroyNotify      result_free_func,
+                               GAsyncReadyCallback callback,
+                               gpointer            user_data)
+{
+  AllTasksResultsClosure *closure = g_new0 (AllTasksResultsClosure, 1);
+
+  closure->results = g_ptr_array_new_with_free_func (result_free_func);
+  closure->remaining = 0;
+  closure->task = g_task_new (NULL, NULL, callback, user_data);
+
+  return closure;
+}
+
+static void
+all_tasks_results_closure_free (AllTasksResultsClosure *closure)
+{
+  g_clear_pointer (&closure->results, g_ptr_array_unref);
+  g_clear_pointer (&closure->task, g_object_unref);
+
+  g_free (closure);
+}
+
+static guint
+all_tasks_results_closure_allocate_new_result_slot (AllTasksResultsClosure *closure)
+{
+  guint last_len = closure->results->len;
+  g_ptr_array_set_size (closure->results, closure->results->len + 1);
+  ++closure->remaining;
+
+  return last_len;
+}
+
+static void
+all_tasks_results_register_result (AllTasksResultsClosure *closure,
+                                   GAsyncResult           *result,
+                                   guint                   index)
+{
+  g_return_if_fail (closure->remaining > 0);
+
+  g_ptr_array_index (closure->results, index) = g_object_ref (result);
```

Comment:
Need to take ownership of the result here so that it can be propagated later (otherwise `g_task_run_in_thread` will unref it when the thread completes)